### PR TITLE
feat: Add async method

### LINKED ISSUE #781: Add support for async methods
We have nice support for `AsyncProperty`s, that can be accessed by async with `.load()` and sync with a `.get`, but if an entity has a method that takes params, currently our convention is a probably-too-cute / verbose `this: Loaded<...>` overlaod:

```ts
  booksToSell(
    this: Loaded<Author, "books">,
    publishedAfter: Date,
  ): Book[];
  booksToSell(
    this: Author,
    publishedAfter: Date,
  ): Promise<Book[]>;
  booksToSell(
    publishedAfter: Date,
  ): MaybePromise<Book[]> {
    return maybePopulateThen(this as Author, "books", (a) => {
      return a.books.get.filter((b) =>b.publishedAt >= publisherAfter);
    });
  }
```

Maybe something like:

```ts
class Author {
  booksToSell = hasAsyncMethod(
    "books",
    [Date],
    (a, publishedAfter) => a.books.get.filter(b => b.publishedAt >= publishedAfter
  );
}

// accessed if loaded
console.log(a.booksToSell.get(publishedAfter));
// accessed if unloaded 
console.log(await a.booksToSell.load(publishedAfter));
```

### PR DESCRIPTION
**Pull request by stephenh**
Fixes #781

### CODE REVIEW

**Review by zgavin — APPROVED** [review]
neat, cute solution 👍 certainly will be nice skip writing a bunch of overloads and having to export a hint

File: packages/orm/src/relations/hasAsyncMethod.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+  get(...args: A): V {
+    if (!this.loaded) {
+      throw new Error("hasAsyncMethod.get was called but not loaded");
+    }
+    return this.fn(this.#entity as Loaded<T, H>, ...args);
+  }
+
+  get isLoaded() {
+    return this.loaded;
```
Comment (zgavin) [inline packages/orm/src/relations/hasAsyncMethod.ts:63]: suggestion: mimic `hasOneDerived`?
```suggestion
    return this.loaded || isLoaded(this.#entity, this.#hint);
```

Comment (stephenh) [reply packages/orm/src/relations/hasAsyncMethod.ts:63]: Ah yep, will ad

**Pull request merged**